Set KEEPALIVE and NO_DELAY socket options on peer connections

alexlarsson opened the pull request.

alexlarsson committed:
Add bus utility to set NO_DELAY and KEEPALIVE socket options

Signed-off-by: Alexander Larsson <alexl@redhat.com>

alexlarsson committed:
Set KEEPALIVE and NO_DELAY socket option on peer connection.

We want this to ensure we detect tcp connection being broken, and
to minimize latency reporting events.

Signed-off-by: Alexander Larsson <alexl@redhat.com>

rhatdan commented on src/agent/agent.c:
```diff
@@ -1029,6 +1029,16 @@ bool agent_start(Agent *agent) {
+                hirte_log_warn("Failed to set NO_DELAY on socket");
+        }
+
+        r = bus_socket_set_no_delay(agent->peer_dbus);
```
This looks like you have a cut and paste error? You set no delay twice?

pypingou replied:
```suggestion
        r = bus_socket_set_keepalive(agent->peer_dbus);
```

rhatdan commented on src/manager/manager.c:
```diff
@@ -367,6 +368,15 @@ static int manager_accept_node_connection(
+        if (r < 0) {
+                hirte_log_warn("Failed to set NO_DELAY on socket");
+        }
+        r = bus_socket_set_no_delay(dbus_server);
```
Cut and paste error again.

pypingou replied:
```suggestion
        r = bus_socket_set_keepalive(dbus_server);
```

sdunnagan commented on src/libhirte/bus/utils.c:
```diff
@@ -164,3 +167,33 @@ char *bus_path_escape(const char *s) {
+        }
+
+        int flag = 1;
+        int r = setsockopt(fd, SOL_SOCKET, SO_KEEPALIVE, (char *) &flag, sizeof(int));
```
TCP keepalives have a default of 2 hours.  On my Fedora host:
```
$ cat /proc/sys/net/ipv4/tcp_keepalive_time
7200
```
Keepalive time can be set for all TCP sockets on the system:
```
$ echo 10 > /proc/sys/net/ipv4/tcp_keepalive_time
```

But that is for all sockets, system wide, which may not be what we want.

I read that there is a ```TCP_KEEPIDLE``` option for setting the keepalive interval
only for the current socket.  So after setting the socket with ```SO_KEEPALIVE```,
we could next use ```TCP_KEEPIDLE``` to set the interval to for example 10 seconds:
```
  int keepidle = 10;
  setsockopt(sock, IPPROTO_TCP, TCP_KEEPIDLE, &keepidle, sizeof(int));
```

alexlarsson replied:
I set KEEPIDLE and KEEPINTVL both to 10.

alexlarsson committed:
Also set KEEPIDLE and KEEPINTVL for KEEPALIVE sockets

We set both of these to 10 sec, because we want to quickly know if
something goes wrong with the network.

Signed-off-by: Alexander Larsson <alexl@redhat.com>
